### PR TITLE
Support utc offset when casting string to utc_datetime

### PR DESCRIPTION
**Pull request by mbuhot**
Currently `utc_datetime` fields silently drops any offset present in a string when `cast`. This PR allows `utc_datetime` to support the same inputs as `DateTime.from_iso8601`.

Since `utc_datetime` also allows for strings without an offset, it will fall back to `NaiveDateTime.from_iso8601` then then convert to `DateTime` in `Etc/UTC` timezone.

The motivation for this PR is a bug that we experienced in our web application, where we are using `utc_datetime` with `embedded_schema` to validate and cast incoming query params:

```elixir
defmodule MyApp.ReportRequest do
  use Ecto.Schema
  alias Ecto.Changeset

  @primary_key false
  embedded_schema do
    field :start_date, :utc_datetime
    field :end_date, :utc_datetime
    field :customer_id, :string
  end

  def validate(params) do
    changeset = 
      %MyApp.ReportRequest{}
      |> Changeset.cast(params, [:start_date, :end_date, :customer_id])
      |> Changeset.validate_required([:start_date, :end_date, :customer_id])

    if changeset.valid? do 
      {:ok, Changeset.apply_changes(changeset)} 
    else 
      {:error, changeset}
    end
  end
end
```

Usage: 

```elixir
iex(2)> MyApp.ReportRequest.validate(%{
  "start_date" => "2001-01-01T12:00:00+10:00", 
  "end_date" => "2010-01-01T12:00:00+10:00", 
  "customer_id" => "abc"})
{:ok,
 %MyApp.ReportRequest{customer_id: "abc",
  end_date: %DateTime{calendar: Calendar.ISO, day: 1, hour: 12,
   microsecond: {0, 0}, minute: 0, month: 1, second: 0, std_offset: 0,
   time_zone: "Etc/UTC", utc_offset: 0, year: 2010, zone_abbr: "UTC"},
  start_date: %DateTime{calendar: Calendar.ISO, day: 1, hour: 12,
   microsecond: {0, 0}, minute: 0, month: 1, second: 0, std_offset: 0,
   time_zone: "Etc/UTC", utc_offset: 0, year: 2001, zone_abbr: "UTC"}}}
```

Notice that it silently dropped the `+10:00` offsets while converting to UTC.

With the changes in this PR, the new output is:

```elixir
iex(2)> MyApp.ReportRequest.validate(%{
  "start_date" => "2001-01-01T12:00:00+10:00", 
  "end_date" => "2010-01-01T12:00:00+10:00", 
  "customer_id" => "abc"})
{:ok,
 %MyApp.ReportRequest{customer_id: "abc",
  end_date: %DateTime{calendar: Calendar.ISO, day: 1, hour: 2,
   microsecond: {0, 0}, minute: 0, month: 1, second: 0, std_offset: 0,
   time_zone: "Etc/UTC", utc_offset: 0, year: 2010, zone_abbr: "UTC"},
  start_date: %DateTime{calendar: Calendar.ISO, day: 1, hour: 2,
   microsecond: {0, 0}, minute: 0, month: 1, second: 0, std_offset: 0,
   time_zone: "Etc/UTC", utc_offset: 0, year: 2001, zone_abbr: "UTC"}}}
```

### CODE REVIEW

**Comment by josevalim** [comment]
This looks good to me. @lau, what are your thoughts? Should this be fixed here or in Elixir?

**Comment by mbuhot** [comment]
@josevalim looks like `DateTime.from_iso8601` isn't available in elixir 1.3.2 :(

Is it too soon to bump the minimum elixir version for ecto?
Alternatively could this be conditionally defined?

```elixir
if function_exported?(DateTime, :from_iso8601, 1) do 
... 
end
```

**Comment by josevalim** [comment]
Yes, we can bump the Elixir requirement as this will be in 2.2 only and there is still some time until we get there.

File: lib/ecto/type.ex [hunk 1]
```diff
@@ -799,6 +799,17 @@ defmodule Ecto.Type do
...
+        case cast_naive_datetime(binary) do
+          {:ok, naive_datetime} -> {:ok, DateTime.from_naive!(naive_datetime, "Etc/UTC")}
+          :error -> :error
+        end
+      {:error, _} -> :error
+    end
+  end
   defp cast_utc_datetime(value) do
     case cast_naive_datetime(value) do
       {:ok, %NaiveDateTime{year: year, month: month, day: day,
```
Comment (josevalim) [inline lib/ecto/type.ex]: We can probably change this clause to use `DateTime.from_naive!` as well, since you are using it above.

**Comment by lau** [comment]
If there is an offset it isn't UTC. The `utc_datetime` type is for storing UTC datetimes. Maybe it should refuse to save strings with offsets if that offset it not 0. If you want to convert it to UTC you can parse it and convert it to UTC before saving it.

Aren't non-UTC `DateTime` structs already refused if they are not "Etc/UTC"? Why should it be different for strings?

Imagine that you have a bunch of strings on different dates and times and hours in different timezones. If you save them to the database and it automatically converts it to UTC all of that information is lost.

2009-12-31T21:00:00-03:00 and 2010-01-01T00:00:00Z are  different years, months, hours, but would be saved to be the same.

Maybe a an error message mentioning a function to transform the string to UTC if that is what is wanted.

**Comment by mbuhot** [comment]
@lau 

> If there is an offset it isn't UTC. 

Yes the incomming string isn't UTC - it would be helpful if during the Ecto `cast` process it would be normalized to UTC before storage.

> The utc_datetime type is for storing UTC datetimes. Maybe it should refuse to save strings with offsets if that offset it not 0. 

Forbidding non-utc string inputs would be better than the current behaviour of silently dropping offsets.

> If you want to convert it to UTC you can parse it and convert it to UTC before saving it.

Can't ecto do this for me? Particularly for the use-case of embedded schemas used to validate and cast API requests (not for DB storage) this would be very useful.

> Aren't non-UTC DateTime structs already refused if they are not "Etc/UTC"? Why should it be different for strings?

No, we've found that even parsing the string into a `DateTime` before passing to ecto, it will silently drop the offset from the struct.

> Imagine that you have a bunch of strings on different dates and times and hours in different timezones. If you save them to the database and it automatically converts it to UTC all of that information is lost.

Yes, that is expected. My columns in the db are declared as `datetime without timezone` and the schema is declared as `:utc_datetime`, so I'm expecting to have UTC whenever values are cast or read from the database.

> 2009-12-31T21:00:00-03:00 and 2010-01-01T00:00:00Z are different years, months, hours, but would be saved to be the same.

That's expected.

> Maybe a an error message mentioning a function to transform the string to UTC if that is what is wanted.

Yes, this would be better than the current situation.
I'd prefer if ecto could just normalize the offsets during cast, saving everyone the headache of having to perform this normalization themselves.

**Comment by josevalim** [comment]
@lau the goal for utc_datetime was always to convert from any timezone to UTC. The reason we don't do that yet is because there is no conversion function in Elixir standard library yet. It is something we have already talked about by e-mail and that we can still hopefully include in time for Elixir v1.5.

As you mentioned, there is a huge concern about converting to UTC, especially for future dates, but if we want to keep all of the information you mentioned, then the best option for now is to use Calecto. The goal is eventually to have Calecto types in Ecto as well but we can't do that before we finish the work on Elixir side of things.

To sum up, we should either error or convert to UTC. I tried to argue above that converting to UTC is the best way to go, because that's what the utc_datetime type advertises. We still have to introduce the raw `:datetime` that will store everything correctly in the database without loss of information (or use Calecto).

Thanks everyone for the discussion! :heart:

**Comment by mbuhot** [comment]
Thanks for the guidance @josevalim.

As @lau mentions, we should consider what the behaviour is when casting a `%DateTime{}` struct that is already in a non-UTC timezone.

I've added a commit that will normalise these `DateTime` to UTC by round-tripping via a unix timestamp. Maybe there is a more direct way?

The tests also show usage of casting maps with string and atom keys to `:utc_datetime`:

```elixir
    assert Ecto.Type.cast(:utc_datetime, %{"year" => 2015, "month" => 1, "day" => 23,
                                           "hour" => 23, "minute" => 50, "second" => 07,
                                           "microsecond" => 8_000}) ==
           {:ok, @datetime_usec}
```

Is this for backwards compatibility with code that was previously using [Ecto.DateTime.cast/1](https://hexdocs.pm/ecto/Ecto.DateTime.html#cast/1) ?

If so, then since `Ecto.DateTime` didn't perform any timezone conversion on these maps, then I don't think `:utc_datetime` would need to either.

**Comment by mbuhot** [comment]
Thanks for the guidance @josevalim.

As @lau mentions, we should consider what the behaviour is when casting a `%DateTime{}` struct that is already in a non-UTC timezone.

I've added a commit that will normalise these `DateTime` to UTC by round-tripping via a unix timestamp. Maybe there is a more direct way?

The tests also show usage of casting maps with string and atom keys to `:utc_datetime`:

```elixir
    assert Ecto.Type.cast(:utc_datetime, %{"year" => 2015, "month" => 1, "day" => 23,
                                           "hour" => 23, "minute" => 50, "second" => 07,
                                           "microsecond" => 8_000}) ==
           {:ok, @datetime_usec}
```

Is this for backwards compatibility with code that was previously using [Ecto.DateTime.cast/1](https://hexdocs.pm/ecto/Ecto.DateTime.html#cast/1) ?

If so, then since `Ecto.DateTime` didn't perform any timezone conversion on these maps, then I don't think `:utc_datetime` would need to either.

**Comment by josevalim** [comment]
> As @lau mentions, we should consider what the behaviour should be when casting a %DateTime{} struct that is already in a non-UTC timezone.

In order to do this, we need to consult the TZ database. The goal is to have this functionality provided as part of Elixir. Round-tripping around Unix is going to fail because Elixir currently only converts to Unix for Etc/UTC. There is nothing we can do here besides waiting for this feature to be added to Elixir or use Calecto.

`Ecto.DateTime` did not know about timezones and therefore cannot be used as a comparison.

**Comment by josevalim** [comment]
Actually, my last comment was incorrect. It seems the to_unix conversion can be done by any timezone, which means we can always go from a timezone to UTC with the information in the struct without the TZ database. @lau can you please confirm? If that's the case, then I believe we should always convert to UTC, as that is the goal of `:utc_datetime`.

**Comment by mbuhot** [comment]
> `Ecto.DateTime` did not know about timezones and therefore cannot be used as a comparison.

OK, I'm just trying to determine if there needs to be special handling of any other data types beyond String and `%DateTime{}`.

The tests show that maps containing `year,month,day,hour,minute,second` keys (as strings or atoms) are supported. Do these maps ever contain a timezone offset?

Is there a scenario where a `DateTime` struct is somehow converted to a plain map that would then need to be cast back to a `DateTime` without data loss?

**Comment by ericmj** [comment]
@mbuhot Refrain from updating the changelog since it can easily cause conflicts with other PRs.

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Comment by mbuhot** [comment]
Thanks @josevalim 👍
